src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,668 @@
+graphs:
+
+@
+0                         == rightVertex 0
+'swap' 1                    == leftVertex 1
+'swap' 1 + 2                == vertices [1] [2]
+('swap' 1) * 2              == edge 1 2
+('swap' 1) + 2 * ('swap' 3)   == overlay (leftVertex 1) (edge 3 2)
+('swap' 1) * (2 + ('swap' 3)) == connect (leftVertex 1) (vertices [3] [2])
+@
+
+__Note:__ the 'Num' instance does not satisfy several "customary laws" of 'Num',
+which dictate that 'fromInteger' @0@ and 'fromInteger' @1@ should act as
+additive and multiplicative identities, and 'negate' as additive inverse.
+Nevertheless, overloading 'fromInteger', '+' and '*' is very convenient when
+working with algebraic graphs; we hope that in future Haskell's Prelude will
+provide a more fine-grained class hierarchy for algebraic structures, which we
+would be able to utilise without violating any laws.
+
+The 'Show' instance is defined using basic graph construction primitives:
+
+@
+show (empty)                       == "empty"
+show 1                             == "rightVertex 1"
+show ('swap' 2)                      == "leftVertex 2"
+show (1 + 2)                       == "vertices [] [1,2]"
+show ('swap' (1 + 2))                == "vertices [1,2] []"
+show (('swap' 1) * 2)                == "edge 1 2"
+show (('swap' 1) * 2 * ('swap' 3))     == "edges [(1,2),(3,2)]"
+show (('swap' 1) * 2 + ('swap' 3))     == "overlay (leftVertex 3) (edge 1 2)"
+show (('swap' 1) * 2 + ('swap' 3) + 4) == "overlay (vertices [3] [4]) (edge 1 2)"
+@
+
+The 'Eq' instance satisfies all axioms of algebraic graphs:
+
+    * 'overlay' is commutative and associative:
+
+        >       x + y == y + x
+        > x + (y + z) == (x + y) + z
+
+    * 'connect' is commutative, associative and has
+    'empty' as the identity:
+
+        >   x * empty == x
+        >   empty * x == x
+        >       x * y == y * x
+        > x * (y * z) == (x * y) * z
+
+    * 'connect' distributes over
+    'overlay':
+
+        > x * (y + z) == x * y + x * z
+        > (x + y) * z == x * z + y * z
+
+    * 'connect' can be decomposed:
+
+        > x * y * z == x * y + x * z + y * z
+
+    * 'connect' has the same effect as 'overlay' on vertices of one part:
+
+        > (leftVertex x)  * (leftVertex y)  == (leftVertex x)  + (leftVertex y)
+        > (rightVertex x) * (rightVertex y) == (rightVertex x) + (rightVertex y)
+
+The following useful theorems can be proved from the above set of axioms.
+
+    * 'overlay' has 'empty'
+    as the identity and is idempotent:
+
+        > x + empty == x
+        > empty + x == x
+        >     x + x == x
+
+    * Absorption and saturation of 'connect':
+
+        > x * y + x + y == x * y
+        >     x * x * x == x * x
+
+When specifying the time and memory complexity of graph algorithms, /n/ and /m/
+will denote the number of vertices and edges in the graph, respectively. In
+addition, /l/ and /r/ will denote the number of vertices in the left and in the
+right part of graph, respectively.
```

Comment:
Just a note: I wonder if by adding `swap` into the algebra we could get some new interesting axioms and theorems. Something to think about!